### DISCUSSION TITLE
email_archive table ought to store the result of each send attempt, to log errors.

### DISCUSSION THREAD

**Discussion by neekfenwick** [zencart/zencart#6449] [Ideas]
The email_archive table stores emails that were sent via zen_mail(), but if any errors occurred during sending the error is silently swallowed.  It would be preferable to have some visibility of which emails were sent without error, and what errors occurred for those emails that failed to send.

I see that in `zen_mail_archive_write()` there is an attempt to catch errors in `$error_msgs` but the variable is never used.  I think we should add a column to the `email_archive` table and store errors there. 

https://github.com/zencart/zencart/blob/a649f475ceaac04599316c2ee2173ed35c0cfd69/includes/functions/functions_email.php#L554

Is there some historical reason for this omission?

If we are to write errors to email_archive, I presume a `TEXT` datatype might be suitable to allow for unknown length error messages, though in reality they're never going to be huge.  What do people think would be a suitable column to add?

For the record, I ran into a Mail addon by swguy at https://www.zen-cart.com/downloads.php?s=ed9d09d7dc00ba783472ae3b0f249789&do=file&id=101 but haven't had time to look at it yet.  Since errors aren't recorded by core ZC I presume this doesn't either, and would need updating if such a feature is added.

Backstory: We use the Gmail SMTP Relay service to send emails from Zen Cart (it acts a lot like any normal gmail connection) which has a whitelisted set of IP addresses which may send via the relay.  A recent live server migration to a new host meant our IP address changed, but we forgot to update the gmail relay whitelist, and all emails sent silently failed, wrote records of "sent" emails to `email_archive`, with no record of which succeeded and which failed.  Cue angry customers etc. etc.

I know that some parts of ZC attempt to read the return value from `zen_mail()` to display `messageStack` messages about any failure, but really a permanent record built into the `mail_archive` table makes a lot of sense as well.

**Comment by drbyte** [comment]
I support the addition of a `TEXT` column to the `email_archive` table containing the `$error_msgs` content.

**Comment by scottcwilson** [comment]
It does generate a debug log on failure, right?  Doesn't that give you what you need?

**Reply by lat9** [reply]
It does, but pretty-much nobody looks at those.  Not sure if anybody's going to look in the `email_archive` table either, since not everyone enables that.

**Reply by scottcwilson** [reply]
I'm just not positive it's worth using the extra space to log failures.  In my experience, most failures are one of these three: 

- misspelled email
- someone using contact us to mail spam (logs for this are now being suppressed)
- incorrectly configured SMTP Server

None of these have value beyond the time they are resolved.

**Comment by neekfenwick** [comment]
> It does generate a debug log on failure, right? Doesn't that give you what you need?

Now I look, I see it does `error_log` if the `ErrorInfo` is not empty but in my experience it's very useful to have errors tied directly to the database records they relate to.

> I'm just not positive it's worth using the extra space to log failures. In my experience, most failures are one of these three:

TEXT has very small storage requirements. https://mariadb.com/kb/en/data-type-storage-requirements/
BLOB, TEXT	len + 2 bytes

It does say at https://mariadb.com/kb/en/text/ "This means that internally TEXT uses less memory for each open table but instead has the additional overhead that each TEXT object needs to be allocated and freed for each row access (with some caching in between)." so if you do a row operation that involves the TEXT column it will involve some extra allocation, so you just have to be careful with `select *` etc when you don't need all fields.

> * misspelled email
> * someone using contact us to mail spam (logs for this are now being suppressed)
> * incorrectly configured SMTP Server

In my case it's a server side process that runs fairly silently.  The ability to query for "all mails sent to betty@a.com" and see specific error messages alongside each one would be very valuable, at least to me.

I agree it's rare, I've needed this maybe 4 times in 10 years, but when it goes wrong it can be extremely difficult figuring out which attempts failed, why, whether they need to be retried etc. etc.  I'd hate to trawl error logs trying to relate the sometimes rather vague error messages against our order processing records.

*edit*: I just deliberately broke my smtp server configuration entry and got this errorinfo:

"SMTP connect() failed. https://github.com/PHPMailer/PHPMailer/wiki/TroubleshootingSMTP server error: Failed to connect to server Additional SMTP info: php_network_getaddresses: getaddrinfo for foosmtp.gmail.com failed: Name or service not known"

If this were sent to `error_log` it would be the devil's own job figuring out what recipient failed and which email failed to send.  I don't know about you other store owners out there but I often find out about such errors after our staff have had many hours sending a wide variety of both human and machine generated emails to customers.

> None of these have value beyond the time they are resolved.

Agreed, but that time between the error happening and the time that the situation is resolved is what error reporting is all about, no? :)

Incidentally I was involved in pushing to fix PHPMailer to make its error reporting useful in this kind of error situation :) https://github.com/PHPMailer/PHPMailer/issues/2570

**Comment by drbyte** [comment]
> TEXT has very small storage requirements.

True. Archived emails already are the largest storage requirement in ZC, so making them a little larger, at scale during a crisis, isn't creating a problem that doesn't already exist.

I'm concerned that the Email Archive Viewer plugin might beg for significant changes in order to group and categorize by email error strings. But that's another topic, which can be handled independently.

I agree with the point that email failures, while logged-to-file, aren't correlatable to which email content failed to send and to whom.

I think storeowners ought to be diligent about deleting archived failed-to-send records, especially after taking action to correct the problem and re-sending said messages.
And they ought to be diligent about purging email logs at least quarterly.

The part that takes the longest during site backups is the database, in particular the email_archive table. (And some payment gateway logging tables.) The bigger the amount of data stored, the larger the daily storage requirement and the larger the backups and their storage requirements.

Again, that said, I support populating the field in the db.

**Comment by neekfenwick** [comment]
Please see PR #6460 for my first stab at this change.

By the way, why is `smtp.gmail.com` hardcoded in a couple of places in `functions_email.php` when there is the database constant `EMAIL_SMTPAUTH_MAIL_SERVER` for this purpose? In our shop we use `smtp-relay.gmail.com` so I've had to modify core ZC to make this work.

**Reply by drbyte** [reply]
That's legacy from when gmail first started to become an option. It needed specific config settings that people were frequently doing wrong, so we added a "Gmail" selection choice and took care of it from there. 

But if you choose Transport of `smtpauth`, you can specify your `smtp-relay.gmail.com` and set all the other settings, without making any coding changes.

**Comment by neekfenwick** [comment]
> The part that takes the longest during site backups is the database, in particular the email_archive table. (And some payment gateway logging tables.) The bigger the amount of data stored, the larger the daily storage requirement and the larger the backups and their storage requirements.

All good points.  I have a cron job that runs every few months that stores old email_archive records to a gz file and deletes them from the table, this makes our daily live database backup more manageable.  Is this a good case for a similar feature to the admin activity log backup feature?  Caveat, as previously said, email_archive may be disabled in config.

In a quiet moment I'd be happy to look at the email management plugin thing but perhaps swguy would be equally happy to take a look, but this year is going crazy with my day job.  I don't know who he is offhand.  Having never seen the code and not knowing when this change might become mainline ZC maybe it's best handled by him.  At the least, a search/filter feature to select records that had error would be useful, as I read in the readme that it has an email resend feature, so after fixing a config problem you could easily resend emails that failed to send during the downtime.  It's likely that any error that occurs, such as a SMTP hostname or port misconfiguration, would affect a block of email attempts in a specific time period with the same error condition, so fancy filtering doesn't sound super important to me.

**Reply by drbyte** [reply]
swguy is @scottcwilson 

He's been maintaining the last updates to the Email Archive Viewer mod. It was written by Frank Koehl years back, and I updated a bunch of things in it for awhile until it was largely stable. I think Scott's updates have been more related to things clients have requested and relatively easy to incorporate.

My comment that it might be a big job is mostly because I think a lot of the plugin's features have been built in layers independent of each other, and it might be at a stage where adding "search for certain errors" (and probably group-by them, and probably mass-re-send-and-purge) might demand a re-think of the UI etc. 
I dunno. Maybe it's super simple to do. 

At least if it "works" for you out-of-the-box and you can simply add some extra controls to suit your needs as time permits, maybe it's a simple solution-for-now and a way to contribute back to the plugin down the road.

Nobody's saying the plugin needs to accommodate the Error Info stuff ..... it just would make sense to start with that plugin if you were gonna make a UI for looking at failed-to-send messages in the archive.

**Reply by scottcwilson** [reply]
> In a quiet moment I'd be happy to look at the email management plugin thing but perhaps swguy would be equally happy to take a look, but this year is going crazy with my day job. I don't know who he is offhand. Having never seen the code and not knowing when this change might become mainline ZC maybe it's best handled by him. 

No thanks.  If you want to do this, you'd have to drive it.

**Reply by neekfenwick** [reply]
@scottcwilson please see https://github.com/scottcwilson/Email_Archive_Manager/pull/2

**Reply by scottcwilson** [reply]
That column should be in the SQL create/update logic for Zen Cart, not for a plugin, since you'll be adding data to it in the Zen Cart core.

**Reply by neekfenwick** [reply]
Yes, it is.
I just realised I didn't test my rewrite of the query building in the PR, will do more thorough testing tomorrow.

**Reply by neekfenwick** [reply]
@scottcwilson I was short on time last night.  The DDL for adding `email_archive.errorinfo` is in my PR (helped by drbyte) e.g. https://github.com/zencart/zencart/pull/6460/commits/bb30380c1ca8b8d3358424eaf7aca64491ef8847
I provided a one-liner example in this PR because the core ZC PR is not merged yet so you may not have had access to that `ADD COLUMN` statement, required for testing my changes to the Email Archive Manager.
I'll do some more testing now and feedback on the addon Issue ticket.

**Reply by scottcwilson** [reply]
@drbyte has created a new repo for Email Archive Manager and is looking to make more changes.  Coordinate with him on this.

**Reply by neekfenwick** [reply]
@drbyte at https://github.com/zencart/zencart/discussions/6449#discussioncomment-9464781 you said @scottcwilson was maintaining the latest updates to the email manager addon, now he says you have a new repo for it, so it sounds like I've made my PR against the wrong repo.  What's going on?

**Reply by drbyte** [reply]
@neekfenwick Your PR is fine. It's great actually. 👍 

What Scott's referring to is that I started working on converting the plugin to a format that ZC's plugin-manager can handle.

I'll incorporate your updates into that as well.

**Reply by neekfenwick** [reply]
Ah thank you @drbyte I see the change you committed at https://github.com/zencart/email-archive-manager/commit/8c740bbee2fc2ff4488442970d20cd15778a039f, and even a little "create errorinfo column if missing" tweak that means it'll work even on an old ZC database.  It looks like there's quite a bit of code difference between that and teh version that I modified, so I imagine you've had to do some difficult merging to get my changes in.

I think you've actually missed my latest changes made today, in commit https://github.com/scottcwilson/Email_Archive_Manager/commit/1053e6caa05d1aae22ef1eae40ca96c7a14607ad I improved the UI and some internal calculations, put three radio buttons in the filter controls for error info control, and the errorinfo column is optional in the on-page results table.  Am I wrong?  Would you want me to clone this newer repo and look at putting in the lastest changes?

![image](https://github.com/zencart/zencart/assets/266415/5afbf08b-6a31-4d54-96b2-3a0c2fada1a9)

I also note that the SELECT includes `'errorinfo' even if it's not being examined, I left that out on purpose because the docs say that a TEXT column requires more work to materialise and calculate, so it seemed wise to omit it from the SELECT list if not required.

**Comment by torvista** [comment]
It seems a bit odd that there is an email archive in the core, but nothing to manage it.

I did some tweaks to the Archive Manager which ended up orphaned in a fork: https://github.com/Zen4All-nl/Email-Archive-Manager-for-Zen-Cart/pull/1

but I remember it was a well-written plugin...perhaps it should be added to the core first and then these enhancements added.

**Reply by neekfenwick** [reply]
I'm hoping to find time to work on the email plugin and add some very basic "show emails with errors" capability.  If I get around to that, we can look at your modifications.  I can't promise anything right now as my day job is quite demanding, at least the email errors work will help us so I can class it as "work" :)
